### PR TITLE
New calculated variables

### PR DESCRIPTION
**Pull request by pnikiel**

### CODE REVIEW

**Comment by pnikiel** [comment]
Still to do:
- fix copyright headers,
- merge support for separate formula for status evalution (conceptually won't change much)
- double-check atomic aspects of value+status passing

**Comment by ben-farnham** [comment]
So far so good by the way, my tests here
https://its.cern.ch/jira/browse/OPCUA-696

**Comment by pnikiel** [comment]
Hi @ben-farnham ,
I believe that in the docs I state exactly opposite to that, i.e. that calculated variables are only for scalars. Do you remember where did you come across it?

**Comment by ben-farnham** [comment]
You do indeed -I respectfully retract that remark!

**Comment by pnikiel** [comment]
Would you have any remarks on more technical aspects like algorithms chosen, integration, synchronization etc ... ?
Is there anything related to the feature that would not let you sleep calmly?

FYI it also does work on PetaLinux/Yocto.

**Comment by ben-farnham** [comment]
I'm getting there - it's a lot to digest...

File: CalculatedVariables/include/CalculatedVariablesSynchronizer.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class Synchronizer
+{
+public:
+    boost::recursive_mutex& mutex() { return m_mutex; }
+
+private:
+    boost::recursive_mutex m_mutex;
+};
+
+typedef std::shared_ptr<Synchronizer> SharedSynchronizer;
```
[the remark this answers is not in the document]
Comment (pnikiel) [reply CalculatedVariables/include/CalculatedVariablesSynchronizer.h]: Will try.

File: CalculatedVariables/src/ParserVariable.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+std::string ParserVariable::name() const
+{
+    return m_notifyingVariable->nodeId().toString().toUtf8();
+}
+
+void ParserVariable::setValue(double v, State state)
+{
+    if (m_synchronizer)
```
[the remark this answers is not in the document]
Comment (pnikiel) [reply CalculatedVariables/src/ParserVariable.cpp]: Discussed and agreed.

File: CalculatedVariables/include/CalculatedVariablesSynchronizer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnikiel) [reply CalculatedVariables/include/CalculatedVariablesSynchronizer.h]: Done.

**Pull request merged**